src/util.js

```diff
@@ -165,3 +165,95 @@ export function sortCustom(array, order, key) {
       return b - a;
   });
 }
+
+// loop through new array of objects. for each object, find object in old
+// array that matches all compare keys, and transfer specified keys from
+// old object to new
+export function transferObjectProps(
+  oldArray,
+  newArray,
+  compareKeys,
+  transferKeys
+) {
+  for (const newElement of newArray) {
+    for (const oldElement of oldArray) {
+      let matches = true;
+      for (const compareKey of compareKeys) {
+        if (newElement[compareKey] !== oldElement[compareKey]) {
+          matches = false;
+          break;
+        }
+      }
+      if (matches) {
+        for (const transferKey of transferKeys)
+          newElement[transferKey] = oldElement[transferKey];
+      }
+    }
+  }
+
+  return newArray;
+}
+
+// same as transferObjectProps, except arrays are now assumed to contain
+// objects with subKey that contains another array that contains the objects
+// to be transferred
+export function transferQueryProps(
+  oldQueries,
+  newQueries,
+  subKey,
+  compareKeys,
+  transferKeys
+) {
+  for (const newQuery of newQueries) {
+    for (const newElement of newQuery[subKey]) {
+      for (const oldQuery of oldQueries) {
+        for (const oldElement of oldQuery[subKey]) {
+          let matches = true;
+          for (const compareKey of compareKeys) {
+            let compare;
+            if (
+              Array.isArray(newElement[compareKey]) &&
+              Array.isArray(oldElement[compareKey])
+            ) {
+              compare = compareArrays(
+                newElement[compareKey],
+                oldElement[compareKey],
+                true
+              );
+            } else
+              compare = newElement[compareKey] === oldElement[compareKey];
+            if (!compare) {
+              matches = false;
+              break;
+            }
+          }
+          if (matches) {
+            for (const transferKey of transferKeys)
+              newElement[transferKey] = oldElement[transferKey];
+          }
+        }
+      }
+    }
+  }
+
+  return newQueries;
+}
+
+// checks if arrays of numbers or strings are equal
+export function compareArrays(array1, array2, checkReverse) {
+  if (checkReverse) {
+    return (
+      array1.length === array2.length &&
+      array1.every(
+        (value, index) =>
+          value === array2[index] ||
+          array1.reverse().every((value, index) => value === array2[index])
+      )
+    );
+  } else {
+    return (
+      array1.length === array2.length &&
+      array1.every((value, index) => value === array2[index])
+    );
+  }
+}
```

Comment:
When `checkReverse` is true, this function will reverse `array1` in place. See:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reverse
In another word, after this function is invoked, `array1` will be the reverse of the one before the invocation. Is this what you want? If yes, this function can be simplified into:
```
function compare_dhu(array1, array2, checkReverse) {
  if (checkReverse) {
    array1 = array1.reverse();
  }

  return (
    array1.length === array2.length &&
      array1.every((value, index) => value === array2[index])
  );
}
```
(If you don't want to reverse `array` in place, you can create a copy of `array1` and simplify the code in a similar fashion.)

Reply:
I forgot that `reverse()` works in place. I didn't mean for that. In fact, the original array shouldn't be modified, and I'm surprised I didn't notice any glitches from this. Further, I noticed a horrible mistake: I'm checking the reverse of the array within each entry check of the non-reversed array. Perhaps that's why I didn't see any glitches, because it reversed it so many times it cancelled out.

Re: your code suggestion, I need to check both forward and backward. So I can't simplify it that way. Although some other simplification may be possible. But here is my suggested correction, for readability and performance:
``` javascript
// checks if arrays of numbers or strings are equal
export function compareArrays(array1, array2, checkReverse) {
  if (checkReverse) {
    return (
      array1.length === array2.length &&
      (array1.every((value, index) => value === array2[index]) ||
        array1
          .slice()
          .reverse()
          .every((value, index) => value === array2[index]))
    );
  } else {
    return (
      array1.length === array2.length &&
      array1.every((value, index) => value === array2[index])
    );
  }
}
```

(performance might actually matter here a little when dealing with thousands of paths)

Reply:
Thanks for the reply. This function is more complex than I thought. Since you always need to check in regular order, how about this?
```
export function compareArrays(array1, array2, checkReverse) {
  if (array1.length !== array2.length) {
    return false;
  }

  const forwardCheck = array1.every((value, index) => value === array2[index]);
  if (!checkReverse || !forwardCheck) {
    return forwardCheck;
  }

  return array1.slice().reverse().every((value, index) => value === array2[index]);
}
```

Reply:
How about a compromise:

```javascript
export function compareArrays(array1, array2, checkReverse) {
  if (array1.length !== array2.length)
    return false;

  if (!checkReverse)
    return array1.every((value, index) => value === array2[index]);

  return (
    array1.every((value, index) => value === array2[index]) ||
    array1
      .slice()
      .reverse()
      .every((value, index) => value === array2[index])
  );
}
```

Reply:
:+1: